hoodie-client/src/main/java/com/uber/hoodie/table/BulkInsertPartitioner.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2017 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.uber.hoodie.table;
+
+import com.uber.hoodie.common.model.HoodieRecord;
+import com.uber.hoodie.common.model.HoodieRecordPayload;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.TreeMap;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.Partitioner;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.PairFunction;
+import org.apache.spark.storage.StorageLevel;
+import scala.Serializable;
+import scala.Tuple2;
+
+/**
+ * Helper class to repartition input records into at least expected number of output spark partitions. It will give
+ * below guarantees
+ * - Average records per output spark partitions will be almost equal to (#inputRecords / #expectedOutputSparkPartitions).
+ * - Output spark partition will have records from only one hoodie partition.
+ */
+public class BulkInsertPartitioner<T extends HoodieRecordPayload> extends Partitioner {
+    private static Logger logger = LogManager.getLogger(BulkInsertPartitioner.class);
+    private static final int BUCKET_MULTIPLIER = 4;
+    private final int expectedOutputSparkPartitions;
+    private final int numBuckets;
+    private HoodiePartitionMapper hoodiePartitionMapper;
+    private int numOfPartitions = -1;
+
+    private BulkInsertPartitioner(int outputSparkPartitions) {
```

Comment:
Removed the constructor once I add partitioner as an argument. I will have to init it via "repartitionRecords" method. Can you please look at it once again?